### ISSUE TITLE
replacing contract installations creates obscure aliases (installation.produce is excess authority)

### ISSUE THREAD

**Issue by dckc** [0xpatrickdev/psm-proposals#4]
This says the proposal may add or change any entry in `agoricNames.installation`:

```json
  "installation": {
...
    "produce": "makeCoreProposalBehavior"
  },
```

There should be no need to use any contracts other than the ones already in `agoricNames.installation`.

Recall earlier discussion:

 - https://github.com/0xpatrickdev/agoric-vault-collateral-proposal/pull/25

in particular:

 - https://github.com/0xpatrickdev/agoric-vault-collateral-proposal/pull/25/commits/7f9ea62f01d636df65852e2bc0fe8ab8efe8f367 chore: prune permits to change agoricNames.installation

**Comment by dckc** [comment]
[agoricNames.installation on emerynet](https://p2p-org.github.io/p2p-agoric-vstorage-viewer/#https://emerynet.rpc.agoric.net|published,published.agoricNames|published.agoricNames.installation) was evidently  updated in block 2133333
at 2023-10-25T22:36:45.458252354Z .

So it looks like this proposal does indeed exercise this authority.

**Comment by 0xpatrickdev** [comment]
Good catch, thank you. We should check this change into the sdk proposal builder for the future. 

> [agoricNames.installation on emerynet](https://p2p-org.github.io/p2p-agoric-vstorage-viewer/#https://emerynet.rpc.agoric.net%7Cpublished,published.agoricNames%7Cpublished.agoricNames.installation) was evidently updated in block 2133333 at 2023-10-25T22:36:45.458252354Z .
> 
> So it looks like this proposal does indeed exercise this authority.

I see `agoricNames.installation` is updated with the same (?) value here, and understand excess authority needs to be avoided. Can you help me understand the impact a bit more for this particular case?

**Comment by dckc** [comment]
> Can you help me understand the impact ...?

good question.

I don't expect any near-term, end-user-visible impact.

More like: it's a bit of technical debt we carry. Some time in the future, we may try to compare installation handles across invitations to PSMs and find that they differ and get confused.

Identity of installation handles is supposed to facilitate things like "that's a good contract" good-housekeeping seal of approval sorts of things.

**Comment by dckc** [comment]
p.s. another impact that came up in discussion with @warner and co is: the old installation handle will become much harder to find. Only the new installation handle will be available via `agoricNames.installation`.

**Comment by 0xpatrickdev** [comment]
Got it, thank you. So instead of just a new instance, we will also have a second installation. And the installation handle for the prior instances of the psm contract will be overridden/cleared from the namehub, which is not desirable.

**Comment by dckc** [comment]
another thing that came up in discussion: [Proposal #17 to add a DAI PSM](https://agoric.explorers.guru/proposal/17) did not include `installation.produce` in its permit, which shows that it's possible to add such an instance without that permission.

```
  "installation": {
    "consume": {
      "contractGovernor": true,
      "psm": true,
      "mintHolder": true
    }
  },
```

Oh... and another observation: this process of adding a PSM is just way too complex/fragile.
One can imagine a form-based tool where you provide an IBC denom, a name, and decimals, and push "go". It shouldn't be much more complicated than that. Good job dealing with things in the present primitive state!

**Comment by 0xpatrickdev** [comment]
I have changes up in #5, but still would like more understanding around a few things. Particularly, around `vatAdminSvc` and any other artifacts around installations.

### 1. How does `restoreRef` work? 

I am struggling to find a helpful type, i see `(ref: unknown) => Promise<unknown>` in most places. Does it return an installation? 

When attempting to remove the restoreRef block, I am greeted with `getManifestForPsm: no function` and the CoreEval fails:
<details>
<summary>git diff</summary>

```diff
diff --git a/gov-start-usdt-psm.js b/gov-start-usdt-psm.js
index d320d21..44ad625 100644
--- a/gov-start-usdt-psm.js
+++ b/gov-start-usdt-psm.js
@@ -97,7 +97,7 @@ const overrideManifest = {
   overrideManifest,
   E,
   log = console.info,
-  restoreRef: overrideRestoreRef,
+  restoreRef,
 }) => {
   const { entries, fromEntries } = Object;

@@ -128,19 +128,6 @@ const overrideManifest = {
     } = allPowers;
     const [exportedGetManifest, ...manifestArgs] = getManifestCall;

-    const defaultRestoreRef = async (ref) => {
-      // extract-proposal.js creates these records, and bundleName is
-      // the name under which the bundle was installed into
-      // config.bundles
-      const p = ref.bundleName
-        ? E(vatAdminSvc).getBundleIDByName(ref.bundleName)
-        : ref.bundleID;
-      const bundleID = await p;
-      const label = bundleID.slice(0, 8);
-      return E(zoe).installBundleID(bundleID, label);
-    };
-    const restoreRef = overrideRestoreRef || defaultRestoreRef;
-
     // Get the on-chain installation containing the manifest and behaviors.
     console.info("evaluateBundleCap", {
       manifestBundleRef,
```

</details>

<details>
<summary>error</summary>

```bash
3-10-26T17:23:53.865Z block-manager: block 72840 commit
2023-10-26T17:23:54.870Z block-manager: block 72841 begin
2023-10-26T17:23:54.875Z block-manager: block 72841 commit
2023-10-26T17:23:55.881Z block-manager: block 72842 begin
2023-10-26T17:23:55.899Z SwingSet: vat: v1: evaluateBundleCap { manifestBundleRef: { bundleID: 'b1-4c34c89b707bc8ece5a41e97e6a354081f7ae8a40391f1462848348613dd1218dcce574b3e30901a9825a966cb85bda6a92ba9f9ce9ba325e4c475f9a678b930' }, exportedGetManifest: 'getManifestForPsm', vatAdminSvc: Promise [Promise] {} }
2023-10-26T17:23:56.510Z SwingSet: vat: v1: execute { exportedGetManifest: 'getManifestForPsm', behaviors: [ 'INVITE_PSM_COMMITTEE_MANIFEST', 'PSM_MANIFEST', 'getManifestForPsm', 'getManifestForPsmGovernance', 'inviteCommitteeMembers', 'inviteToEconCharter', 'makeAnchorAsset', 'startEconCharter', 'startPSM' ] }
2023-10-26T17:23:56.511Z SwingSet: vat: v1: CORE_EVAL failed: (TypeError#2)
2023-10-26T17:23:56.511Z SwingSet: vat: v1: TypeError#2: getManifestForPsm: no function
2023-10-26T17:23:56.512Z SwingSet: vat: v1: TypeError: getManifestForPsm: no function
 at getManifestForPsm (.../inter-protocol/src/proposals/startPSM.js:488)
 at behavior (#8665:153)
 at ()
 at ()

2023-10-26T17:23:56.512Z SwingSet: xsnap: v1: TypeError#2 ERROR_NOTE: Sent as error:liveSlots:v1#70002
2023-10-26T17:23:56.513Z SwingSet: ls: v1: Logging sent error stack (TypeError#2)
2023-10-26T17:23:56.525Z SwingSet: xsnap: v10: UnhandledPromiseRejectionWarning: (RemoteTypeError(error:liveSlots:v1#70002)#5)
2023-10-26T17:23:56.526Z SwingSet: xsnap: v10: RemoteTypeError(error:liveSlots:v1#70002)#5: getManifestForPsm: no function
2023-10-26T17:23:56.526Z SwingSet: xsnap: v10: TypeError: getManifestForPsm: no function
 at makeError (/bundled-source/.../node_modules/ses/src/error/assert.js:243)
 at decodeErrorCommon (/bundled-source/.../node_modules/@endo/marshal/src/marshal.js:281)
 at decodeFromSmallcaps (/bundled-source/.../node_modules/@endo/marshal/src/encodeToSmallcaps.js:434)
 at fromCapData (/bundled-source/.../node_modules/@endo/marshal/src/marshal.js:356)
 at notifyOnePromise (/bundled-source/.../packages/swingset-liveslots/src/liveslots.js:1225)
 at notify (/bundled-source/.../packages/swingset-liveslots/src/liveslots.js:1249)
 at dispatchToUserspace (/bundled-source/.../packages/swingset-liveslots/src/liveslots.js:1514)
 at runWithoutMetering (/bundled-source/.../packages/swingset-xsnap-supervisor/lib/supervisor-subprocess-xsnap.js:60)
 at ()

2023-10-26T17:23:56.532Z block-manager: block 72842 commit
2023-10-26T17:23:56.907Z block-manager: block 72843 begin
2023-10-26T17:23:56.911Z block-manager: block 72843 commit
2023-10-26T17:23:57.916Z block-manager: block 72844 begin
2023-10-26T17:23:57.919Z block-manager: block 72844 commit
```
</details>

Is a better approach (to eliminate reliance on `vatAdminSvc`), maybe using agoricNames as a lookup for the instance?


### 2. Do my proposes changes in #5 prevent multiple installs? 

https://github.com/0xpatrickdev/psm-proposals/blob/2023-10-25-agoric-3-USDC/gov-start-usdc-psm.js#L141

Here, install is still being called. Will we now have the opposite effect on `agoricNames.installation`, where the second installation is used and the ref of the new PSM is not present on the NameHub?


### 3. `vatAdminSvc` in bundleCap, bundleCapFlow

I am still working through this, but this is what I think I understand so far. 
 - `bundleCap` is `EndoZipBase64Bundle`, or `({moduleFormat: "endoZipBase64"; endoZipBase64: string; endoZipBase64Sha512: string; })`
 - `evaluateBundleCap` (reliant on `importBundle` from `endo`) decodes the bundleCap, are seemingly returns references to installed modules in a compartment (`manifestNS`)
 
https://github.com/0xpatrickdev/psm-proposals/blob/2023-10-25-agoric-3-USDC/gov-start-usdc-psm.js#L165-L172

This piece, I am struggling to understand. I could not locate a type for `manifestNS` (it is unclear to me what `options` looks like), and am not sure I understand `exportedGetManifest` . It seems like it may be grabbing a permit/permissions, or that already happens early in `evaluateBundleCap`. 


If there is a different way to approach this, without `vatAdminSvc`, I'd be interested to hear!

**Comment by dckc** [comment]
I wondered whether the 6 PSMs started by the vaults launch each created their own installation handle for psm (b1-c25fb) and mintHolder (b1-b183e). @warner did some digging and confirmed that this is the case:

```
v9.vs.vom.o+d22/14|{"bundleLabel":{"body":"#\"b1-b183e\"","slots":[]}}
v9.vs.vom.o+d22/16|{"bundleLabel":{"body":"#\"b1-b183e\"","slots":[]}}
v9.vs.vom.o+d22/18|{"bundleLabel":{"body":"#\"b1-b183e\"","slots":[]}}
v9.vs.vom.o+d22/20|{"bundleLabel":{"body":"#\"b1-b183e\"","slots":[]}}
v9.vs.vom.o+d22/22|{"bundleLabel":{"body":"#\"b1-b183e\"","slots":[]}}
v9.vs.vom.o+d22/24|{"bundleLabel":{"body":"#\"b1-b183e\"","slots":[]}}
v9.vs.vom.o+d22/13|{"bundleLabel":{"body":"#\"b1-c25fb\"","slots":[]}}
v9.vs.vom.o+d22/15|{"bundleLabel":{"body":"#\"b1-c25fb\"","slots":[]}}
v9.vs.vom.o+d22/17|{"bundleLabel":{"body":"#\"b1-c25fb\"","slots":[]}}
v9.vs.vom.o+d22/19|{"bundleLabel":{"body":"#\"b1-c25fb\"","slots":[]}}
v9.vs.vom.o+d22/21|{"bundleLabel":{"body":"#\"b1-c25fb\"","slots":[]}}
v9.vs.vom.o+d22/23|{"bundleLabel":{"body":"#\"b1-c25fb\"","slots":[]}}
```

**Comment by dckc** [comment]
> If there is a different way to approach this, without vatAdminSvc, I'd be interested to hear!

Not really... not without a wholesale change of approach.

It's used to get the `startPSM.js` code.

**Comment by dckc** [comment]
fixed in #5
